### PR TITLE
Upgrading to PyLCA v2.0

### PR DESCRIPTION
**Pull request by TJTapajyoti**
This branch updates the PyLCA to v2 with faster calculations, no optimization, better debugged USLCI database and overall better results.

### CODE REVIEW

File: celavi/pylca_celavi/insitu_emission.py [hunk 1]
```diff
@@ -79,56 +78,16 @@ def solver_optimization(tech_matrix,F,process, df_with_all_other_flows):
...
-
-    def supply_rule(model, i):
-      return sum(model.x[i, j] * model.s[j] for j in model.j) >= model.f[i]
-    model.supply = Constraint(model.i, rule=supply_rule, doc='Equations')
-
-    def objective_rule(model):
-      return sum(model.s[j] for j in model.j)
-    model.objective = Objective(rule=objective_rule, sense=minimize, doc='Define objective function')
+    tm= tech_matrix.to_numpy()
+    det = np.linalg.det(tm)
```
Comment (rjhanes) [inline celavi/pylca_celavi/insitu_emission.py:85]: det is currently unused

File: celavi/pylca_celavi/insitu_emission.py [hunk 2]
```diff
@@ -198,15 +157,15 @@ def runner(tech_matrix,F,yr,i,j,k,final_demand_scaler,process,df_with_all_other_
...
        res.loc[:,'year'] =  yr
        res.loc[:,'facility_id'] = i
        res.loc[:,'stage'] = j
        res.loc[:,'material'] = k
     else:        
-       print(f"Insitu emission optimization failed for {k} at {j} in {yr}")
+       print(f"Insitu emission calculation failed for {k} at {j} in {yr}")
     
     res = electricity_corrector_before20(res)
     res['route_id'] = None
```
Comment (rjhanes) [inline celavi/pylca_celavi/insitu_emission.py:171]: The `route_id` parameter can be removed from the function definition and the docstring

File: celavi/pylca_celavi/insitu_emission.py [hunk 3]
```diff
@@ -267,6 +226,10 @@ def model_celavi_lci_insitu(f_d, yr, fac_id, stage, material, df_emissions):
     final_dem = product_df.merge(f_d, left_on=0, right_on='flow name', how='left')
```
Comment (rjhanes) [inline celavi/pylca_celavi/insitu_emission.py:226]: For line 214: `final_lci_result` is not used

File: celavi/pylca_celavi/insitu_emission.py [hunk 4]
```diff
@@ -267,6 +226,10 @@ def model_celavi_lci_insitu(f_d, yr, fac_id, stage, material, df_emissions):
...
     final_dem = final_dem.fillna(0)
     chksum = np.sum(final_dem['flow quantity'])
+    if chksum == 0:
```
Comment (rjhanes) [inline celavi/pylca_celavi/insitu_emission.py:229]: Lines 229-233 can be combined in a single if/else statement

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 5]
```diff
@@ -2,18 +2,19 @@
 import pandas as pd
 import numpy as np
 import time
+import sys
 from pyomo.environ import ConcreteModel, Set, Param, Var, Constraint, Objective, minimize, SolverFactory
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:6]: Line 6 can be deleted as none of the imports appear to be used

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 6]
```diff
@@ -2,18 +2,19 @@
 import pandas as pd
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:2]: On line 1: the pickle import is not used

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 7]
```diff
@@ -2,18 +2,19 @@
...
 import time
+import sys
 from pyomo.environ import ConcreteModel, Set, Param, Var, Constraint, Objective, minimize, SolverFactory
 import pyutilib.subprocess.GlobalData
 pyutilib.subprocess.GlobalData.DEFINE_SIGNAL_HANDLERS_DEFAULT = False
 
 
-def model_celavi_lci_background(f_d, yr, fac_id, stage,material, route_id, uslci_filename,
+def model_celavi_lci_background(f_d, yr, fac_id, stage,material, route_id, uslci_tech_filename,uslci_emission_filename,uslci_process_filename,
                                 lci_activity_locations):
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:12]: `lci_activity_locations` is not used

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 8]
```diff
@@ -29,8 +30,13 @@ def model_celavi_lci_background(f_d, yr, fac_id, stage,material, route_id, uslci
       material of LCA analysis
     route_id: str
         Unique identifier for transportation route.
-    uslci_filename: str
+    uslci_tech_filename: str
       filename for the USLCI inventory
+    uslci_process_filename: str
+      filename for the USLCI process
+    uslci_emission_filename: str
+      filename for the USLCI emissions
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:38]: Can you expand on these docstrings (either here or wherever it makes the most sense) - Looks like "tech" is the technology matrix and the other two are lists of names? Give some idea of the file content.

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 9]
```diff
@@ -41,264 +47,11 @@ def model_celavi_lci_background(f_d, yr, fac_id, stage,material, route_id, uslci
        Final LCA results in the form of a dataframe after performing after calculation checks
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:47]: Here too, give some idea of the content of the returned DataFrame (emissions or indicators? column names?)

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 10]
```diff
@@ -308,92 +61,37 @@ def solver_optimization(tech_matrix,F):
...
+        process_emissions: str
+             filename for the emissions dataframe
         
         Returns
         -------
         pd.DataFrame
             LCA results
         """
+        tm = tech_matrix.to_numpy()
+        det = np.linalg.det(tm)
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:73]: det is not used

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 11]
```diff
@@ -308,92 +61,37 @@ def solver_optimization(tech_matrix,F):
...
-        #All emission are grouped and summed up together in this dataframe
-        product_results_total = product_results_df.groupby(by = ['conjoined_flownames','unit'])['value'].agg(sum).reset_index()
-    
-        return emissions_results_total,product_results_total
-            
+        final_result = scv_df.merge(process_emissions, on = ['process'])
+        final_result['total_emission'] = final_result['scaling vector'] * final_result['value']
+        final_result['value'] = final_result['total_emission']
+        em_result = final_result[['product','value','process','unit']]
+        co2 = em_result[em_result['product'] == 'Carbon dioxide']
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:85]: co2 is not used

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 9 above)
Comment (rjhanes) [reply celavi/pylca_celavi/pylca_opt_background.py:47]: I won't leave a comment everywhere, but in general the return values for these functions should have more information

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 12]
```diff
@@ -308,92 +61,37 @@ def solver_optimization(tech_matrix,F):
...
+        emissions_results_total = final_result.groupby(['product','unit'])['value'].agg('sum').reset_index()
+      
+        return emissions_results_total
+
     
-    def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
+    def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler,process_emissions):
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:91]: Minor quibble, doesn't really need to be fixed - There's three different functions in pyLCIA called "runner". Can they be re-named for greater clarity? Can there be just one runner function used throughout pyLCIA?

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 13]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
         """
         tim0 = time.time()
         res = pd.DataFrame()
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:123]: Line 123 can be deleted - in the case that solver fails to find results, it should return an empty DataFrame

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 14]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
...
         if not res.empty:
           res.loc[:,'year'] =  i
           res.loc[:,'facility_id'] =  l
           res.loc[:,'stage'] = j
           res.loc[:,'material'] = k
-          res.loc[:, 'route_id'] = route_id
+          res.loc[:,'route_id'] = route_id
     
-          print(str(i) +' - '+j + ' - ' + k)
+          print(str(i) +' - '+j + ' - ' + k, flush = True)
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:133]: Do we need this print statement? Looks like there's something printed if the calculations work and something printed if they fail.

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 15]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
...
-          res2.loc[:, 'route_id'] = route_id
-    
-        else:
-           print(f"optimization pylca-opt-background emission failed for {k} at {j} in {i}")
-           pass
+          print(f"pylca-opt-background emission failed for {k} at {j} in {i}", flush = True) 
+          pass       
+        
     
         print(str(time.time() - tim0) + ' ' + 'taken to do this run',flush=True)
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:140]: Do we always need the timing printed out? Can there be a `verbose` parameter added that controls how much information is printed?

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 16]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
...
+    f_d = f_d.merge(process_adder, left_on = "flow name", right_on = "product")
+    #This list of products and processes essentially help to determine the indexes and the products and processes
+    #to which they belong. 
+    uslci_products = pd.DataFrame(tech_matrix.index)
+    uslci_process = list(tech_matrix.columns)
+
+
     #Have to edit this final demand to match the results from CELAVI
-    f_d['flow name'] = f_d['flow name'] +'@' + f_d['flow name']
+    #f_d['flow name'] = f_d['flow name'] +'@' + f_d['flow name']
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:159]: (throughout) Remove commented-out old code

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 17]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
...
 
-    final_dem = uslci_product_df.merge(f_d, left_on=0, right_on='flow name', how='left')
-    final_dem = final_dem.fillna(0)
     chksum = np.sum(final_dem['flow quantity'])
-    #To make the optimization easier
+    if chksum == 0:
+        print('Final demand construction failed. No value. csv file for error checking created. /n Check the intermediate demand file and final demand check file')
+        final_dem.to_csv('../../celavi-data/generated/Final_demand_check_file.csv')
+        uslci_products.to_csv('../../celavi-data/generated/uslciproducts_file.csv')
+        f_d.to_csv('../../celavi-data/generated/demandforeground.csv')
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:182]: Lines 180-182: Remove all hard-coded file directories and names, replace with parameters in the YAML file(s)

File: celavi/pylca_celavi/pylca_opt_background.py [hunk 18]
```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
...
     f_d = f_d.sort_values(['year'])
 
 
     #Replace electricity    
-    f_d  = f_d .replace(to_replace='electricity@electricity', value='electricity, at grid@electricity, at grid, us, 2010')
+    #f_d  = f_d .replace(to_replace='electricity@electricity', value='electricity, at grid, us, 2010@electricity, at grid, us, 2010')
+    #f_d['flow name'] = f_d['flow name'] +'@' + f_d['location']
+    #f_d['flow name'] = f_d['flow name'].str.lower()
+    f_d['conjoined_flownames'] = f_d['conjoined_flownames'].str.lower()  
+    f_d.to_csv('../../celavi-data/generated/intermediate_demand_foreground_uslci_data.csv', mode = 'a')
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_background.py:170]: This line currently breaks the code - see my comment on the main PR page

File: celavi/pylca_celavi/pylca_opt_foreground.py [hunk 19]
```diff
@@ -7,10 +7,7 @@
 import os
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_foreground.py:7]: The warnings, sys, multiprocessing, time, and os imports are not used

**Review by akey7 — DISMISSED** [review]
Please see my comment about the hardcoded pathname in the source code and how it crashes the run on my machine.

File: celavi/pylca_celavi/pylca_opt_foreground.py [hunk 20]
```diff
@@ -98,54 +95,13 @@ def solver_optimization(tech_matrix,F,process, df_with_all_other_flows):
...
-    model.objective = Objective(rule=objective_rule, sense=minimize, doc='Define objective function')
-
-    def pyomo_postprocess(options=None, instance=None, results=None):
-        df = pd.DataFrame.from_dict(model.s.extract_values(), orient='index', columns=[str(model.s)])
-        return df
-
-    # This is an optional code path that allows the script to be run outside of
-    # pyomo command-line.  For example:  python transport.py
+    tm= tech_matrix.to_numpy()
+    det = np.linalg.det(tm)
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_foreground.py:99]: det is unused

File: celavi/pylca_celavi/pylca_opt_foreground.py [hunk 21]
```diff
@@ -231,15 +188,15 @@ def runner(tech_matrix,F,yr,i,j,k,route_id,state,final_demand_scaler,process,df_
...
 
 
 def model_celavi_lci(f_d,yr,fac_id,stage,material,route_id,state,df_static,dynamic_lci_filename,electricity_grid_spatial_level,intermediate_demand_filename):
 
     """
-    Main function of this module which received information from DES interface and runs the suppoeting optimization functions. 
+    Main function of this module which received information from DES interface and runs the supporting LCA functions. 
     Creates the technology matrix and the final demand vector based on input data. 
-    Performs necessary checks before and after the LCA optimization calculation. 
+    Performs necessary checks before and after the LCA calculation.
```
Comment (rjhanes) [inline celavi/pylca_celavi/pylca_opt_foreground.py:199]: More detail here would be helpful - what checks are performed, what errors do they prevent?

**Review by rjhanes — CHANGES_REQUESTED** [review]
I left comments on specific lines with the recommended changes

File: celavi/pylca_celavi/insitu_emission.py (hunk 1 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/insitu_emission.py:85]: Resolved

File: celavi/pylca_celavi/insitu_emission.py (hunk 2 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/insitu_emission.py:171]: Removed

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 14 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:133]: I kept this print statement in order to track the solutions as they occur in real time. I will say once we see that code is running perfectly on exploring the results, we can delete this line.

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 15 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:140]: A verbose parameter would be a good way to stop these print statements (also for the previous comment). One question, is there an already used verbose parameters in the code that I can pull from __main__ or scenario or should I create one within the LCA environment? @rjhanes

File: celavi/pylca_celavi/pylca_opt_foreground.py (hunk 21 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_foreground.py:199]: Added detailed information

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 15 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:140]: Added verbose to LCA

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 14 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:133]: Verbose parameter added.

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 12 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:91]: Rectified with new names.

File: celavi/pylca_celavi/pylca_opt_background.py (hunk 9 above)
Comment (TJTapajyoti) [reply celavi/pylca_celavi/pylca_opt_background.py:47]: I have added detailed information about the returned dataframes for most functions. Some functions are still left. I will keep on adding them. But other than that all other corrections have been completed and the code is running. I will keep an issue open for the final editing for docstrings for me.

**Comment by TJTapajyoti** [comment]
I have added detailed information about the returned dataframes from the functions inside the pylcia.py files for most functions. Some functions are still left. I will keep on adding them. But other than that all other corrections have been completed and the code is running. I will keep an issue open for the final editing for docstrings for me to resolve. However, if you want you can proceed with the merge and conflict resolution. @rjhanes

**Review by rjhanes — APPROVED** [review]
I added a commit that will allow the LCIA verbosity to be set through the case study YAML file like the cost graph verbosity parameter. I tested on the pylca2.0-upgrade data branch, with and without shortcut calculations, and setting the verbosity parameter to 0 and 1, and saw no problems. It is indeed a good 3-4 orders of magnitude faster per LCIA calculation - nice job, TJ! 

Next step here is for me to merge in my stochastic uncertainty branch to this pylcav2.0-upgrade branch for a quick test & debug session. I'll send out a notice once that's done and we can merge this branch into dev.

**Comment by rjhanes** [comment]
Following PR 170 being merged into the `pylcav2.0-upgrade` branch, here is the testing & debug status (being updated as I complete steps):

- [x] Test on tiny data using baseline MCiC-MC case with no uncertainty - Complete, no errors, results are as expected
- [ ] Test on tiny data using 10_fgic_oaat.yaml file, 8/16 version on OneDrive
- [ ] Test on tiny data using stochastic-scenario.yaml file, 8/16 version on OneDrive

**Comment by TJTapajyoti** [comment]
> Following PR 170 being merged into the `pylcav2.0-upgrade` branch, here is the testing & debug status (being updated as I complete steps):
> 
>     * [x]  Test on tiny data using baseline MCiC-MC case with no uncertainty - Complete, no errors, results are as expected (~200 s)
> 
>     * [x]  Test on tiny data using 10_fgic_oaat.yaml file, 8/16 version on OneDrive - Complete, all good (636 s)
> 
>     * [x]  Test on tiny data using stochastic-scenario.yaml file, 8/16 version on OneDrive - Complete, all good (756 s _without_ a starting shortcut database)
> 
> 
> Testing is complete, works great! @TJTapajyoti I think you're OK to merge into `develop` whenever you're ready. ~I'll work on updating the data repo branches in the meantime.~ All data repo branches should now work with the pyLCA2.0/stochastic updates.
> 
> For HPC runs, there is a scenario YAML file on OneDrive, under uncertainty/stochastic-yamls, called stochastic-scenario-4states. This will execute 1,000 runs varying the 8 parameters we discussed the other day. If it causes any problems, let me know as usual and I'll work to correct it ASAP.

Thanks. I will get to this by tomorrow. I need to push the changes to the warning statements. Will push and perform a check so that it works fine.

**Pull request merged**